dimod/reference/composites/cutvertex.py

```diff
@@ -0,0 +1,281 @@
+# Copyright 2019 D-Wave Systems Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+# =============================================================================
+"""
+A composite that breaks the problem into sub-problems corresponding to the
+biconnected components of the binary quadratic model graph before sending to its child sampler.
+"""
+
+from dimod.sampleset import as_samples
+from dimod.core.composite import ComposedSampler
+from dimod.sampleset import SampleSet
+import dimod
+import networkx as nx
```

Comment:
See discussion in https://github.com/dwavesystems/dimod/pull/539#discussion_r332116231. However, in this case the functions are potentially more involved, so rewriting might be more trouble than it's worth.

Reply:
The nx.articulation_points and nx.biconnected_components aren't too difficult actually, and pure python. It's basically just DFS. So if you'd rather have dimod.articulation_points and dimod.biconnected_components we could do that (although it does seem like a lot of code replication).

Reply:
Yeah exactly, https://github.com/dwavesystems/dimod/commit/e5dd5f39c374209b1a7c843c5f42a3687fa563d2 added bfs so dfs as well is not such a stretch, but at some point we end up just re-implementing NetworkX and there's no point in that. On the other hand, casting BQMs to NetworkX graphs has a non-trivial copy cost so working natively on BQMs can be nice :man_shrugging:.

Reply:
IMHO, we should try to use nx's function until it's clear there is a drastic slowdown because of copy costs or pure Python. 

Glad to learn about `nx.articulation_points`. I somehow missed the `articulation_points` function last summer when I was working on this problem. That would have simplified things. 👍

Reply:
I tend to agree. It's just frustrating because the BQM and networkx graph are so close in API, for a lot of nx functions the only thing blocking them working on the BQM directly is a decorator checking whether it's a directed graph or not. We could potentially wrap the BQM with some attributes to get around that but I am sure there would be edge cases.

Reply:
BTW, I've had great results using Metis in the past to decompose graphs. IIRC, it has a Python interface, but would add another requirement (including C backend 👎 )

Reply:
pymetis, I think. Yeah metis is great (although overkill if you just want cut vertices rather a partition of an arbitrary graph). In principle we could extend the functionality in CutVertexComposite to cut sets larger than a single vertex, but the run time grows exponentially with the size of the cut set.